cdap-app-templates/cdap-etl/cdap-data-streams-base/src/main/java/io/cdap/cdap/datastreams/SparkStreamingPipelineRunner.java

```diff
@@ -185,4 +214,124 @@ protected SparkCollection<Object> handleJoin(Map<String, SparkCollection<Object>
+
+          //process the remaining stages
+          for (int i = 1; i < topologicalOrder.size(); i++) {
+            String stageName = topologicalOrder.get(i);
+            processStage(phaseSpec, sourcePluginType, sec, stagePartitions, pluginContext, collectors, pipelinePhase,
+                         functionCacheFactory, macroEvaluator, emittedRecords, groupedDag, groups, branchers, shufflers,
+                         sinkRunnables, stageName, time.milliseconds(), context);
+          }
+
+          //We should have all the sink runnables at this point, execute them
+          executeSinkRunnables(sec, sinkRunnables);
+
+          if (dStream instanceof StreamingEventHandler) {
+            ((StreamingEventHandler) dStream).onBatchCompleted(streamingContext);
+          } else if (dStream.dstream() instanceof StreamingEventHandler) {
+            ((StreamingEventHandler) (dStream.dstream())).onBatchCompleted(streamingContext);
+          }
+
+        }, Exception.class);
+
+      });
+
+    } else {
+      super.processDag(phaseSpec, sourcePluginType, sec, stagePartitions, pluginContext, collectors, pipelinePhase,
+                       functionCacheFactory, macroEvaluator, groupedDag, groups, branchers, shufflers);
+    }
+  }
+
+  @Override
+  protected Runnable getSparkSinkRunnable(StageSpec stageSpec, SparkCollection<Object> stageData,
+                                          SparkSink<Object> sparkSink, long time,
+                                          @Nullable DatasetContext datasetContext) throws Exception {
+    if (nativeStateStoreEnabled) {
+      return () -> {
+        try {
+          new StreamingSparkSinkFunction<>(sec, stageSpec, datasetContext)
+            .call(stageData.getUnderlying(), Time.apply(time));
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      };
+    }
+    return super.getSparkSinkRunnable(stageSpec, stageData, sparkSink, time, datasetContext);
+  }
+
+  @Override
+  protected Runnable getBatchSinkRunnable(FunctionCache.Factory functionCacheFactory, StageSpec stageSpec,
+                                          SparkCollection<Object> stageData,
+                                          PluginFunctionContext pluginFunctionContext, long time,
+                                          @Nullable DatasetContext datasetContext) {
+    if (nativeStateStoreEnabled) {
+      return () -> {
+        try {
+          new StreamingBatchSinkFunction<>(sec, stageSpec, functionCacheFactory.newCache(), datasetContext).call(
+            stageData.getUnderlying(), Time.apply(time));
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      };
+    }
+    return super.getBatchSinkRunnable(functionCacheFactory, stageSpec, stageData, pluginFunctionContext, time,
+                                      datasetContext);
+  }
+
+  @Override
+  protected Runnable getGroupSinkRunnable(PhaseSpec phaseSpec, Set<String> groupStages,
```

Comment:
similar comment about inverting the check

Reply:
Done